pulp_gem/specs.py

```diff
@@ -60,8 +140,30 @@ def analyse_gem(file_obj):
     """
     with TarFile(fileobj=file_obj) as archive:
         with archive.extractfile("metadata.gz") as md_file:
-            data = yaml.safe_load(gzip.decompress(md_file.read()))
+            data = yaml.load(gzip.decompress(md_file.read()), Loader=RubyMarshalYamlLoader)
+    gem_info = {
+        "name": data.values["name"],
+        "version": data.values["version"].values["version"],
+    }
+    # Sanitize name
+    if not NAME_REGEX.match(gem_info["name"]):
+        raise ValueError(f"Invalid gem name: {gem_info['name']}")
+    # Sanitize version
+    if VERSION_REGEX.match(gem_info["version"]):
+        gem_info["prerelease"] = False
+    elif PRERELEASE_VERSION_REGEX.match(gem_info["version"]):
+        gem_info["prerelease"] = True
+    else:
+        raise ValueError(f"Invalid version string: {gem_info['version']}")
+    for key in ("required_ruby_version", "required_rubygems_version"):
+        if (requirement := data.values.get(key)) is not None:
+            gem_info[key] = _collapse_requirement(requirement)
```

Comment:
Why use `_collapse_requirement` here and not above when reading the info file?

Reply:
This is translating the version requirement for a single package into a string like `">= 2.0.1&< 3.0"` from the weird ruby-marshal format to be used in the plain text that is the info file (well, here we store it on the model to be used at publish time...). When reading that info-file it is already "collapsed".